pkg/controller/ps/controller.go

```diff
@@ -457,6 +460,57 @@ func (r *PerconaServerMySQLReconciler) doReconcile(
 	return nil
 }
 
+func (r *PerconaServerMySQLReconciler) validate(ctx context.Context, cr *apiv1alpha1.PerconaServerMySQL) error {
+	if err := validateClusterType(ctx, r.Client, cr); err != nil {
+		return errors.Wrap(err, "validate cluster type")
+	}
+	return nil
+}
+
+func validateClusterType(ctx context.Context, cl client.Client, cr *apiv1alpha1.PerconaServerMySQL) error {
+	sts := new(appsv1.StatefulSet)
+	if err := cl.Get(ctx, types.NamespacedName{Name: mysql.Name(cr), Namespace: cr.Namespace}, sts); err != nil {
+		if k8serrors.IsNotFound(err) {
+			return nil
+		}
+		return errors.Wrap(err, "failed to get mysql sts")
+	}
+
+	containerName := mysql.ComponentName
+	var container *corev1.Container
+	for _, c := range sts.Spec.Template.Spec.Containers {
+		if c.Name != containerName {
+			continue
+		}
+
+		container = &c
+		break
+	}
+	if container == nil {
+		return errors.New("failed to get mysql container")
+	}
+
+	clusterType := ""
+	for _, e := range container.Env {
+		if e.Name != naming.EnvMySQLClusterType {
+			continue
+		}
+
+		clusterType = e.Value
+		break
+	}
+
+	if clusterType == "" {
+		return errors.New("failed to get mysql cluster type")
+	}
+
+	if cr.Spec.MySQL.ClusterType == apiv1alpha1.ClusterType(clusterType) {
+		return nil
+	}
+
+	return errors.Errorf("wrong mysql cluster type, expected: %s", clusterType)
```

Comment:
I think we can let the users know somehow that they are not suppossed to try to change the cluster type, in a more direct manner through this log.

I propose this message: `return errors.Errorf("cluster type cannot be changed from %s to %s on a running cluster", currentClusterType, cr.Spec.MySQL.ClusterType)`

Reply:
https://github.com/percona/percona-server-mysql-operator/pull/939/commits/76bb5f956bf603287d629f57f977ff4b7d0ea812